eclair-core/src/main/scala/fr/acinq/eclair/package.scala

```diff
@@ -159,4 +159,54 @@ package object eclair {
     * We use this in the context of timestamp filtering, when we don't need an upper bound.
     */
   val MaxEpochSeconds = Duration.fromNanos(Long.MaxValue).toSeconds
+
+  /**
+    * One MilliSatoshi is a thousand of a Satoshi, the smallest unit usable in bitcoin
+    * @param amount
+    */
+  case class MilliSatoshi(amount: Long) {
+    // @formatter:off
+    def toLong = amount
+    def +(other: MilliSatoshi) = MilliSatoshi(amount + other.amount)
+    def -(other: MilliSatoshi) = MilliSatoshi(amount - other.amount)
+    def *(m: Long) = MilliSatoshi(amount * m)
+    def /(d: Long) = MilliSatoshi(amount / d)
+    def compare(other: MilliSatoshi): Int = if (amount == other.amount) 0 else if (amount < other.amount) -1 else 1
+    def <= (that: MilliSatoshi): Boolean = compare(that) <= 0
+    def >= (that: MilliSatoshi): Boolean = compare(that) >= 0
+    def <  (that: MilliSatoshi): Boolean = compare(that) < 0
+    def >  (that: MilliSatoshi): Boolean = compare(that) > 0
+    def unary_-() = MilliSatoshi(-amount)
+    def truncateToSatoshi: Satoshi = Satoshi(amount / 1000)
+    // @formatter:on
+  }
+
+  implicit class ToMilliSatoshiConversion(amount: BtcAmount) {
+    def toMilliSatoshi: MilliSatoshi = amount match {
+      case sat: Satoshi => satoshi2millisatoshi(sat)
+      case millis: MilliBtc => satoshi2millisatoshi(millibtc2satoshi(millis))
+      case bitcoin: Btc => satoshi2millisatoshi(btc2satoshi(bitcoin))
+    }
+  }
+
+  implicit object NumericMilliSatoshi extends Numeric[MilliSatoshi] {
+    override def plus(x: MilliSatoshi, y: MilliSatoshi): MilliSatoshi = MilliSatoshi(x.amount + y.amount)
+    override def minus(x: MilliSatoshi, y: MilliSatoshi): MilliSatoshi = MilliSatoshi(x.amount - y.amount)
+    override def times(x: MilliSatoshi, y: MilliSatoshi): MilliSatoshi = MilliSatoshi(x.amount * y.amount)
+    override def negate(x: MilliSatoshi): MilliSatoshi = MilliSatoshi(-x.amount)
+    override def fromInt(x: Int): MilliSatoshi = MilliSatoshi(x)
+    override def toInt(x: MilliSatoshi): Int = x.toLong.toInt
+    override def toLong(x: MilliSatoshi): Long = x.toLong
+    override def toFloat(x: MilliSatoshi): Float = x.toLong
+    override def toDouble(x: MilliSatoshi): Double = x.toLong
+    override def compare(x: MilliSatoshi, y: MilliSatoshi): Int = x.compare(y)
+  }
+
+  private def satoshi2millisatoshi(input: Satoshi): MilliSatoshi = MilliSatoshi(input.amount * 1000L)
+
+  def maxOf(x: MilliSatoshi, y: MilliSatoshi) = MilliSatoshi(Math.max(x.amount, y.amount))
```

Comment:
Somehow i find more explicit `Math.max` since using the standard(java) library functions enables the reader to assume the implementation is correct, it's not a big deal to change it though.